### PR TITLE
Add gettext support to the android version

### PR DESCRIPTION
**Pull request by oleastre**
Add compact, short information about your PR for easier understanding:

- Goal of the PR
Add gettext and internationalization support in the android version

- How does the PR work?
Add dependency on a static version of gettext, and add mo files to the packaged android assets.
Depends on https://github.com/minetest/minetest_android_deps_binaries/pull/3

- Does it resolve any reported issue?
Solves #6093

## To do

This PR is Ready for Review.

## How to test

Build and launch an android client, set the language to something not english in the settings, restart and enjoy a translated mintest.

### CODE REVIEW

**Comment by MoNTE48** [comment]
Implementation of MultiCraft. 
https://github.com/MultiCraft/MultiCraft/tree/master/lib/intl
Very minimal version of the library. It may be more rational to use it on Android, instead of a full (and pervasive) gettext.
It can also be used on other platforms for guaranteed support for locales.

**Comment by oleastre** [comment]
Note that this PR does not automatically detect the android language. 
I could add this in a further PR by adding a JNI binding to fetch the info from the java side of the code.

**Comment by MoNTE48** [comment]
> I could add this in a further PR by adding a JNI binding to fetch the info from the java side of the code.

What for? JNI can detect language, just read the documentation.

**Comment by oleastre** [comment]
As stated in the binary part of the PR, I used gettext, as it's already the translation engine for minetest. I also considered https://github.com/tinygettext/tinygettext and https://github.com/j-jorge/libintl-lite, but since the `runtime` part of gettext is small, self contained, and easy to compile using the NDK, I thought it would be more straight forward to use it to enable translations in the android client.

**Comment by oleastre** [comment]
> What for? JNI can detect language, just read the documentation.

That was the purpose of my comment, you have to add a JNI call to get the current locale. 
Basically one need to all `Locale.getDefault()` in java and forward that to the gettext initialization code.
Or did I miss something in the doc ?

**Comment by MoNTE48** [comment]
@oleastre yes, miss it:
https://github.com/MultiCraft/MultiCraft/blob/7059d450f24843ca4b7699a651eaf131e84cd76c/src/defaultsettings.cpp#L483

**Comment by oleastre** [comment]
> @oleastre yes, miss it:
Thanks, I'm not a full tie android developer and most of the time, I only play with the java side and almost never touch the NDK. 
So adding that call is quite easy. I'll add to this PR.

File: build/android/app/build.gradle [hunk 1]
```diff
@@ -75,10 +75,13 @@ task prepareAssets() {
...
-		// ToDo: fix broken locales
-		from "${projRoot}/po" into "${assetsFolder}/po"
-	}*/
+	fileTree("${projRoot}/po").include("**/*.po").forEach { poFile ->
+		def moPath = "${assetsFolder}/locale/${poFile.parentFile.name}/LC_MESSAGES/"
+		file(moPath).mkdirs()
+		exec {
+			commandLine 'msgfmt', '-o', "${moPath}/minetest.mo", poFile
+		}
+	}
```
Comment (MoNTE48) [inline build/android/app/build.gradle:84]: There should be a comment that indicates that gradle will not be able to do this on Windows at all, as well as on macOS without `brew install gettext`!

File: build/android/native/jni/Android.mk [hunk 2]
```diff
@@ -76,6 +81,8 @@ LOCAL_CFLAGS += \
 	-DUSE_FREETYPE=1                \
 	-DUSE_LEVELDB=0                 \
 	-DUSE_LUAJIT=1                  \
+	-DUSE_GETTEXT=1                 \
+	-DRUN_IN_PLACE=0                \
```
Comment (MoNTE48) [inline build/android/native/jni/Android.mk:85]: Why is this needed?

File: src/gui/modalMenu.cpp [hunk 3]
```diff
@@ -140,7 +140,8 @@ bool GUIModalMenu::preprocessEvent(const SEvent &event)
 			std::string label = wide_to_utf8(getLabelByID(hovered->getID()));
 			if (label.empty())
 				label = "text";
-			message += gettext(label) + ":";
+			message += gettext(label.c_str());
+			message += ":";
```
Comment (MoNTE48) [inline src/gui/modalMenu.cpp:144]: Why?

File: build/android/native/jni/Android.mk (hunk 2 above)
Comment (oleastre) [reply build/android/native/jni/Android.mk:85]: When testing I got a locale path set to ` ../build/locale`, I tried to remove it with this.
But after that I added the `path_locale` config in `porting_android`, so maybe this nit needed any more. Can try to remove it.

Comment (MoNTE48) [reply build/android/native/jni/Android.mk:85]: When setting all paths to porting_android this line does nothing

File: src/gui/modalMenu.cpp (hunk 3 above)
Comment (oleastre) [reply src/gui/modalMenu.cpp:144]: `message` is of type `std::string`, and the type returned by `gettext(...)` is `char*`. 
And since we are adding `":"` just after, the compiler was considering that constant also as `char*`. But adding two `char*` is not supported. 
The conversion to `std::string` comes with the `+=` assignation
So this was the simplest way to just force the conversion of the two values to `std::string`.
Note that it was not needed before because `gettext` was just a macro returning the original type that was `std::string`.

File: build/android/app/build.gradle (hunk 1 above)
Comment (oleastre) [reply build/android/app/build.gradle:84]: True. Is there a place to add  this note ?

Comment (MoNTE48) [reply build/android/app/build.gradle:84]: @sfan5 your suggestions?

File: src/gui/modalMenu.cpp (hunk 3 above)
Comment (MoNTE48) [reply src/gui/modalMenu.cpp:144]: Ok, but then what is wrong with `gettext(label) + ":"`?

Comment (oleastre) [reply src/gui/modalMenu.cpp:144]: that's where the problem is: we try to do: `std::string += ( char* + char* )` the compiler tries to solve the left operand before the assignation. So, it tries to sum two char pointers.

Comment (oleastre) [reply src/gui/modalMenu.cpp:144]: I could have written `std::string(gettext(label)) + std::string(":")`, but it's probably less readable than a new line ;)

File: build/android/native/jni/Android.mk (hunk 2 above)
Comment (oleastre) [reply build/android/native/jni/Android.mk]: Solved.

**Comment by oleastre** [comment]
I added automatic locale detection if it's not set in the UI or minetest.conf

**Comment by oleastre** [comment]
Rebased on latest master after 5.3.0 release

File: build/android/app/build.gradle (hunk 1 above)
Comment (SmallJoker) [reply build/android/app/build.gradle:83]: Windows can do that as long `msgfmt` is within the path. All it needs is a comment stating that this section requires `gettext` to be installed and available in PATH.

**Comment by SmallJoker** [comment]
I don't have an Android setup, but the code looks good. :+1: if someone else can confirm this as working.

File: build/android/app/build.gradle (hunk 1 above)
Comment (oleastre) [reply build/android/app/build.gradle:83]: I'll see if I can add a test and break early with a meaningful error message.

File: src/gettext.cpp [hunk 4]
```diff
@@ -202,7 +207,12 @@ void init_gettext(const char *path, const std::string &configured_language,
 #endif // ifndef _WIN32
 	}
 	else {
+#ifdef __ANDROID__
+		char lang[3] = {0};
 		 /* set current system default locale */
```
Comment (MoNTE48) [inline src/gettext.cpp:212]: Move this before #ifdef

File: build/android/native/jni/Android.mk [hunk 5]
```diff
@@ -59,6 +59,11 @@ include $(PREBUILT_STATIC_LIBRARY)
 #LOCAL_SRC_FILES := deps/Android/OpenSSL/${NDK_TOOLCHAIN_VERSION}/$(APP_ABI)/libcrypto.a
 #include $(PREBUILT_STATIC_LIBRARY)
 
+include $(CLEAR_VARS)
+LOCAL_MODULE := GetText
```
Comment (MoNTE48) [inline build/android/native/jni/Android.mk:63]: Just `gettext` as here https://www.gnu.org/software/gettext/

**Comment by MoNTE48** [comment]
I can confirm that this completely works and has been using MultiCraft for a while.
I support this PR!

**Comment by oleastre** [comment]
I'll take time this week to address the few comments, I've left this pending for too much time

**Comment by MoNTE48** [comment]
@rubenwardy this really can be merged. Tested and used in my fork

**Comment by srifqi** [comment]
I can confirm that this PR works, too.

**Comment by AntumDeluge** [comment]
Is there a reason this hasn't been merged yet?

**Comment by Pevernow** [comment]
@oleastre Are you still work in it?
If not,I would like to make a new pull request for it to rebase it.

**Comment by oleastre** [comment]
@Pevernow I did not updated it for a while, even locally. 
If you want to rebase and create a new PR, I'm happy with it; otherwise, I'll have to re-install the full android stack to do it and test against the latest minetest versions

**Comment by SmallJoker** [comment]
Rebased/continued in #11435
